presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java

```diff
@@ -262,103 +230,70 @@ LogicalPropertiesImpl build()
-            equivalenceClassProperty = equivalenceClassProperty.project(inverseVariableAssignments);
-            normalizeKeyPropertyAndSetMaxCard(keyProperty, maxCardProperty, equivalenceClassProperty);
-            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
-        }
+        //project both equivalence classes and key property
+        Map<VariableReferenceExpression, VariableReferenceExpression> inverseVariableAssignments = inverseVariableAssignments(assignments);
+        keyProperty = keyProperty.project(new InverseVariableMappingsWithEquivalence(equivalenceClassProperty, inverseVariableAssignments));
+        equivalenceClassProperty = equivalenceClassProperty.project(inverseVariableAssignments);
+        normalizeKeyPropertyAndSetMaxCard(keyProperty, maxCardProperty, equivalenceClassProperty);
+
+        return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
     }
 
     /**
-     * This logical properties builder should be used by PlanNode that propagate their
+     * This logical properties builder should be used by PlanNode's that propagate their
      * source properties and add a limit. For example, TopNNode and LimitNode.
      */
-    public static class PropagateAndLimitBuilder
-    {
-        private final LogicalPropertiesImpl sourceProperties;
-        private final long limit;
-        private final MaxCardProperty maxCardProperty = new MaxCardProperty();
-        private final EquivalenceClassProperty equivalenceClassProperty;
-        private final KeyProperty keyProperty = new KeyProperty();
-
-        PropagateAndLimitBuilder(LogicalPropertiesImpl sourceProperties, long limit, FunctionResolution functionResolution)
-        {
-            requireNonNull(sourceProperties, "sourceProperties is null");
-            this.sourceProperties = sourceProperties;
-            this.limit = limit;
-            this.equivalenceClassProperty = new EquivalenceClassProperty(functionResolution);
-        }
 
-        LogicalPropertiesImpl build()
-        {
-            keyProperty.addKeys(sourceProperties.keyProperty);
-            maxCardProperty.update(sourceProperties.maxCardProperty);
-            maxCardProperty.update(limit);
-            equivalenceClassProperty.update(sourceProperties.equivalenceClassProperty);
-            if (maxCardProperty.isAtMostOne()) {
-                keyProperty.empty();
-            }
-            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
+    public static LogicalPropertiesImpl propagateAndLimitProperties(LogicalPropertiesImpl sourceProperties,
+                                                                    long limit,
+                                                                    FunctionResolution functionResolution)
+    {
+        requireNonNull(sourceProperties, "sourceProperties is null");
```

Comment:
we usually only do null checks in constructors.

also usually only use the final keyword for class variables.

(these comments apply throughout this class)